src/pairs.rs

```diff
@@ -0,0 +1,326 @@
+            infix,
+            suffix: "",
+        }
+    }
+}
+
+// Flattens a tree of pairs into a list and tries to rewrite them all at once.
+// FIXME would be nice to reuse the lists API for this, but because each separator
+// can be different, we can't.
+pub(crate) fn rewrite_all_pairs(
+    expr: &ast::Expr,
+    shape: Shape,
+    context: &RewriteContext,
+) -> Option<String> {
+    // First we try formatting on one line.
+    if let Some(list) = expr.flatten(context, false) {
+        if let Some(r) = rewrite_pairs_one_line(&list, shape, context) {
+            return Some(r);
+        }
+    }
+
+    // We can't format on line, so try many. When we flatten here we make sure
+    // to only flatten pairs with the same operator, that way we don't
+    // necessarily need one line per sub-expression, but we don't do anything
+    // too funny wrt precedence.
+    expr.flatten(context, true)
+        .and_then(|list| rewrite_pairs_multiline(list, shape, context))
+}
+
+// This may return a multi-line result since we allow the last expression to go
+// multiline in a 'single line' formatting.
+fn rewrite_pairs_one_line<T: Rewrite>(
+    list: &PairList<T>,
+    shape: Shape,
+    context: &RewriteContext,
+) -> Option<String> {
+    assert!(list.list.len() >= 2, "Not a pair?");
+
+    let mut result = String::new();
+    let base_shape = shape.block();
+
+    for (e, s) in list.list[..list.list.len()]
+        .iter()
+        .zip(list.separators.iter())
+    {
+        let cur_shape = base_shape.offset_left(last_line_width(&result))?;
+        let rewrite = e.rewrite(context, cur_shape)?;
+
+        if !is_single_line(&rewrite) || result.len() > shape.width {
+            return None;
+        }
+
+        result.push_str(&rewrite);
+        result.push(' ');
+        result.push_str(s);
+        result.push(' ');
+    }
+
+    let last = list.list.last().unwrap();
+    let cur_shape = base_shape.offset_left(last_line_width(&result))?;
+    let rewrite = last.rewrite(context, cur_shape)?;
+    result.push_str(&rewrite);
+
+    if first_line_width(&result) > shape.width {
+        return None;
+    }
+
+    // Check the last expression in the list. We let this expression go over
+    // multiple lines, but we check that if this is necessary, then we can't
+    // do better using multi-line formatting.
+    if !is_single_line(&result) {
+        let multiline_shape = shape.offset_left(list.separators.last().unwrap().len() + 1)?;
+        let multiline_list: PairList<T> = PairList {
+            list: vec![last],
+            separators: vec![],
+            separator_place: list.separator_place,
+        };
+        // Format as if we were multi-line.
+        if let Some(rewrite) = rewrite_pairs_multiline(multiline_list, multiline_shape, context) {
+            // Also, don't let expressions surrounded by parens go multi-line,
+            // this looks really bad.
+            if rewrite.starts_with('(') || is_single_line(&rewrite) {
```

Comment:
so the reason I think it looks bad is because it doesn't respect precedence. If you have `a * (b + c)` where all three are long expressions, I think

```
a
* (b + c)
```
is the best grouping because it reflects the precedence.

```
a
* (b
+ c)
```

is ok, but sub-optimal,

```
a * (b
+ c)
```

is "really bad" because it is misleading wrt precedence (imagining that the expressions are long, it is fairly easy to miss the parens and only see the line breaks.